File: Matrix completion/Python/GNMR_completion.py
import numpy as np
from scipy import sparse
from scipy.sparse import linalg
from sklearn.preprocessing import normalize

# initialization options
INIT_WITH_SVD = 0
INIT_WITH_RANDOM = 1
INIT_WITH_USER_DEFINED = 2

def GNMR_completion(X, omega, rank, verbose=True, alpha=1, max_outer_iter=100,
    max_inner_iter=2000, lsqr_init_tol=1e-15, lsqr_smart_tol=True, lsqr_smart_obj_min=1e-5, lsqr_col_norm=False,
    init_option=INIT_WITH_SVD, init_U=None, init_V=None,
    stop_relRes=1e-16, stop_relDiff = -1, stop_relResDiff = -1,
    normalize_X=False, r_projection_in_iteration=False):
    """
    Run GNMR algorithm for matrix completion
    (in the following, relRes == ||X_hat - X||_F / ||X_hat||_F on the observed entires)
    :param ndarray X: Input matrix (n1,n2). Unobserved entries should be zero
    :param ndarray omega: Mask matrix (n1,n2). 1 on observed entries, 0 on unobserved
    :param int rank: Underlying rank matrix
    :param bool verbose: if True, display intermediate results
    :param int alpha: indicating the variant of GNMR (e.g., 1: setting, 0: averaging, -1: updating)
    :param int max_outer_iter: Maximal number of outer iterations
    :param int max_inner_iter: Maximal number of inner iterations
    :param float lsqr_init_tol: initial tolerance of the LSQR solver
    :param bool lsqr_smart_tol: if True, when relRes <= lsqr_smart_obj_min, use lsqr_tol=objective**2
    :param float lsqr_smart_obj_min: relRes threshold to begin smart tolerance from
    :param bool lsqr_col_norm: normalize columns of the least-squares matrix
    :param int init_option: how to initialize U and V (INIT_WITH_SVD, INIT_WITH_RAND, or INIT_WITH_USER_DEFINED)
    :param ndarray init_U: U initialization (n1,rank), used in case init_option==INIT_WITH_USER_DEFINED
    :param ndarray init_V: V initialization (n2,rank), used in case init_option==INIT_WITH_USER_DEFINED
    :param float stop_relRes: relRes threshold for ealy stopping (relevant to noise-free case), -1 for disabled
    :param float stop_relDiff: relative X_hat difference threshold for ealy stopping, -1 for disabled
    :param float stop_relResDiff: relRes difference difference threshold for early stopping, -1 for disabled
    :param bool normalize_X: work with normalized matrix X, may enhance performance in some cases
    :parma bool r_projection_in_iteration: if true, error estimation at each iteration
      is calculated for the best rank-r approximation using SVD
    :return: GNMR's estimate, final iteration number, convergence flag and all relRes
    """
    n1, n2 = X.shape
    num_visible_entries = np.count_nonzero(omega)
    visible_ratio = num_visible_entries / (n1*n2)

    # initial estimate
    if init_option == INIT_WITH_SVD:
      (U, _, V) = linalg.svds(X, k=rank, tol=1e-16)
      V = V.T
    elif init_option == INIT_WITH_RANDOM:
      U = np.random.randn(n1, rank)
      V = np.random.randn(n2, rank)
      U = np.linalg.qr(U)[0]
      V = np.linalg.qr(V)[0]
    else:
      U = init_U
      V = init_V

    # normalize X
    if normalize_X:
      sv_avg_estimate = (1/visible_ratio) * np.sum(np.linalg.svd(X, compute_uv=False)[1]) / rank
      X = X / sv_avg_estimate

    # generate sparse indices to accelerate future operations
    sparse_matrix_rows, sparse_matrix_columns = generate_sparse_matrix_entries(omega, rank, n1, n2)
    # generate (constant) b for the least squares problem
    b = generate_b(X, omega, n1, n2)

    # before iterations
    early_stopping_flag = False
    relRes = 1
    all_relRes = [relRes]
    best_relRes = np.max(np.abs(X))
    X_hat = U @ V.T
    X_hat_best_2r = X_hat  # stores best intermediate rank 2r estimate

    # iterations
    iter_num = 0
    while iter_num < max_outer_iter and not early_stopping_flag:
        iter_num += 1
        
        # determine LSQR tolerance
        current_tol = lsqr_init_tol
        if lsqr_smart_tol and relRes < lsqr_smart_obj_min:
          current_tol = min(current_tol, relRes**2)
        
        # solve the least squares problem
        A = generate_sparse_A(U, V, omega, sparse_matrix_rows, sparse_matrix_columns, num_visible_entries, n1, n2, rank)
        b = generate_b(X + alpha * U @ V.T, omega, n1, n2)
        if lsqr_col_norm:
          A, scale_vec = rescale_A(A)
        x, _, _, res = linalg.lsqr(A, b, atol=current_tol, btol=current_tol, iter_lim=max_inner_iter)[:4]
        relRes = res / np.linalg.norm(b)
        if lsqr_col_norm:
          x = x * scale_vec
        x = convert_x_representation(x, rank, n1, n2)

        # obtain new estimates for U and V
        U_tilde, V_tilde = get_U_V_from_solution(x, rank, n1, n2)
        U_next = 0.5 * (1 - alpha) * U + U_tilde
        V_next = 0.5 * (1 - alpha) * V + V_tilde
        
        # get new estimate and calculate corresponding error
        X_hat_previous = X_hat
        X_hat_2r = U @ V_next.T + U_next @ V.T - U @ V.T
        if normalize_X:
          X_hat_2r *= sv_avg_estimate
        if r_projection_in_iteration:
          (U_r, Sigma_r, V_r) = linalg.svds(X_hat_2r, k=rank, tol=1e-17)
          X_hat = U_r @ np.diag(Sigma_r) @ V_r
        else:
          X_hat = U_next @ V_next.T
        X_hat_diff =  np.linalg.norm(X_hat - X_hat_previous, ord='fro') / np.linalg.norm(X_hat, ord='fro')

        all_relRes.append(relRes)
        if relRes < best_relRes:
          best_relRes = relRes
          X_hat_best_2r = X_hat_2r

        # update U, V
        U = U_next
        V = V_next

        # report
        if verbose:
          print("[INSIDE GNMR] iter: " + str(iter_num) + ", relRes: " + str(relRes))

        # check early stopping criteria
        if stop_relRes > 0:
          early_stopping_flag |= relRes < stop_relRes
        if stop_relDiff > 0:
          early_stopping_flag |= X_hat_diff < stop_relDiff
        if stop_relResDiff > 0:
          early_stopping_flag |= np.abs(relRes / all_relRes[-2] - 1) < stop_relResDiff
        if verbose and early_stopping_flag:
          print("[INSIDE GNMR] early stopping")

    # return
    convergence_flag = iter_num < max_outer_iter
    (U_r, Sigma_r, V_r) = linalg.svds(X_hat_best_2r, k=rank, tol=1e-17)
    X_hat = U_r @ np.diag(Sigma_r) @ V_r
    return X_hat, iter_num, convergence_flag, all_relRes


# auxiliary functions

def rescale_A(A):
    A = sparse.csc_matrix(A)
    scale_vec = 1. / linalg.norm(A, axis=0)
    return normalize(A, axis=0), scale_vec


def convert_x_representation(x, rank, n1, n2):
    recovered_x = np.array([x[k * rank + i] for i in range(rank) for k in range(n2)])
    recovered_y = np.array([x[rank * n2 + j * rank + i] for i in range(rank) for j in range(n1)])
    return np.append(recovered_x, recovered_y)


def generate_sparse_matrix_entries(omega, rank, n1, n2):
    row_entries = []
    columns_entries = []
    row = 0
    for j in range(n1):
        for k in range(n2):
            if 0 != omega[j][k]:
                # add indices for U entries
                for l in range(rank):
                    columns_entries.append(k * rank + l)
                    row_entries.append(row)
                # add indices for V entries
                for l in range(rank):
                    columns_entries.append((n2 + j) * rank + l)
                    row_entries.append(row)
                row += 1
    return row_entries, columns_entries


def generate_sparse_A(U, V, omega, row_entries, columns_entries, num_visible_entries, n1, n2, rank):
    # we're generating row by row
    data_vector = np.concatenate(
        [np.concatenate([U[j], V[k]]) for j in range(n1) for k in range(n2) if 0 != omega[j][k]])
    return sparse.csr_matrix(sparse.coo_matrix((data_vector, (row_entries, columns_entries)),
                                               shape=(num_visible_entries, rank * (n1 + n2))))

def generate_b(X, omega, n1, n2):
    return np.array([X[j][k] for j in range(n1)
                     for k in range(n2) if 0 != omega[j][k]])


def get_U_V_from_solution(x, rank, n1, n2):
    VT = np.array([x[i * n2:(i + 1) * n2] for i in range(rank)])
    UT = np.array([x[rank * n2 + i * n1: rank * n2 + (i + 1) * n1] for i in range(rank)])
    return UT.T, VT.T


def get_estimated_value(x, U, V, rank, n1, n2):
    # calculate U's contribution
    estimate = np.sum(
        [np.dot(U.T[i].reshape(n1, 1), np.array(x[i * n2:(i + 1) * n2]).reshape(1, n2)) for i in
         range(rank)],
        axis=0)
    # calculate V's contribution
    estimate += np.sum(
        [np.dot(x[rank * n2 + i * n1: rank * n2 + (i + 1) * n1].reshape(n1, 1),
                V.T[i].reshape(1, n2))
         for i in range(rank)], axis=0)
    return estimate

File: Matrix completion/MATLAB/GNMR_completion.m
function [X_hat, all_relRes, iter, convergence_flag] = ...
    GNMR_completion(X,omega,r,opts)
%
% Written by Pini Zilber & Boaz Nadler / 2021
% based on the code of R2RILS by Jonatahan Bauch, Boaz Nadler and Pini Zilber
% with modifications by Christian Kuemmerle (sparse matrix construction of A)
%
% INPUT: 
% X = the observed matrix
% omega = list of pairs (i,j) of the observed entries
% r = target rank of the underlying matrix
% opts = options meta-variable (see opts_default below for details)
%
% OUTPUT:
% X_hat = rank-r approximation of the underlying matrix
% all_relRes = list of the residual errors of the least-squares problem throughout the iterations
% iter = final iteration number
% convergence_flag = indicating whether the algorithm converged
%

%% configurations: default values for option variables
opts_default.verbose = 1;                       % display intermediate results
opts_default.alpha = 1;                         % variant parameter (e.g., 1: setting, 0: averaging, -1: updating)
% number of iterations and tolerance
opts_default.max_outer_iter = 100;              % maximal number of outer iterations
opts_default.max_inner_iter = 2000;             % maximal number of inner iterations for the LSQR solver
opts_default.inner_init_tol = 1e-15;            % initial tolerance of the LSQR solver
opts_default.LSQR_smart_tol = 1;                % use LSQR_tol==relRes^2 when relRes is low enough
opts_default.LSQR_smart_obj_min = 1e-5;         % relRes threshold to start using LSQR smart tol
% initialization
opts_default.init_option = 0;                   % 0 for SVD, 1 for random, 2 for opts.init_U, opts.init_V
opts_default.init_U = NaN;                      % if opts.init_option==2, use this initialization for U
opts_default.init_V = NaN;                      % if opts.init_option==2, use this initialization for V
% early stopping criteria (-1 to disable a criterion)
opts_default.stop_relRes = 1e-16;               % small relRes threshold (relevant to noise-free case)
opts_default.stop_relDiff = -1;                 % small relative X_hat difference threshold
opts_default.stop_relResDiff = -1;              % small relRes difference threshold
opts_default.stop_relResStuck_ratio = -1;       % stop if minimal relRes didn't change by a factor of stop_relResNoChange_ratio...
opts_default.stop_relResStuck_iters = -1;       % ... in the last #stop_relResNoChange_iters outer iterations
% additional configurations
opts_default.normalize_X = 0;                   % work with normalized matrix X, may enhance performance in some cases
opts_default.r_projection_in_iteration = 0;     % if true, error estimation at each iteration is calculated
                                                % for the best rank-r approximation using SVD

% for each unset option set its default value
fn = fieldnames(opts_default);
for k=1:numel(fn)
    if ~isfield(opts,fn{k}) || isempty(opts.(fn{k}))
        opts.(fn{k}) = opts_default.(fn{k});
        fn{k} = opts_default.(fn{k});
    end
end

%% some definitions
[n1, n2] = size(X);   % number of rows / colums
m = size(omega,1);   % number of observed entries
colind_A = generate_sparse_matrix_indices(omega,r,n2); % for the sparse matrix A

%% initialize U and V (of sizes n1 x r and n2 x r)
if opts.init_option == 0
    % initialization by rank-r SVD of observed matrix
    [U, ~, V] = svds(X,r);
elseif opts.init_option == 1
    % initialization by random orthogonal matrices
    Z = randn(n1,r);
    [U, ~, ~] = svd(Z,'econ'); 
    Z = randn(n2,r);
    [V, ~, ~] = svd(Z,'econ'); 
else
    % initiazliation by user-defined matrices
    U = opts.U_init;
    V = opts.V_init; 
end

%% normalize X
if opts.normalize_X
    visible_ratio = m / (n1*n2);
    [~, Sigma, ~] = svds(X);
    sig_avg_est = (1/visible_ratio) * sum(sum(Sigma)) / r; % (check if helps)
    X = X / sig_avg_est;
end

%% before iterations
X_hat_previous = U*V';   % stores rank-r projection of previous intermediate rank 2r estimate
all_relRes = zeros(opts.max_outer_iter,1);
relRes = Inf;
best_relRes = Inf;
convergence_flag = 0;

%% iterations
for iter = 1:opts.max_outer_iter

    %% construct variables for LSQR
    A = generate_sparse_A(U,V,omega,colind_A);
    
    % update rhs (b in ||Ax-b||^2)
    rhs = zeros(m,1);  % vector of visible entries in matrix X
    X_updated = X + opts.alpha * U*V';
    for counter=1:m
        rhs(counter) = X_updated(omega(counter,1),omega(counter,2));
    end

    %% solve LSQR
    % determine tolerance for LSQR solver
    LSQR_tol = opts.inner_init_tol;
    if opts.LSQR_smart_tol && relRes < opts.LSQR_smart_obj_min
        LSQR_tol = min(LSQR_tol, relRes^2);
    end
    LSQR_tol = max(LSQR_tol, 2*eps);  % to supress warning
    % solve the least squares problem
    [Z, ~, relRes, LSQR_iters_done] = lsqr(A,rhs,LSQR_tol,opts.max_inner_iter);
        % LSQR finds the minimum norm solution and is much faster than lsqminnorm

    %% construct Utilde and Vtilde from the solution Z
    Utilde = zeros(size(U));
    Vtilde = zeros(size(V)); 
    nc_list = r * [0:1:(n2-1)]; 
    for i=1:r
        Vtilde(:,i) = Z(i+nc_list); 
    end
    nr_list = r * [0:1:(n1-1)]; 
    start_idx = r*n2; 
    for i=1:r
        Utilde(:,i) = Z(start_idx + i + nr_list);
    end
    
    %% calculate U_next, V_next
    U_next = 0.5*(1 - opts.alpha) * U + Utilde;
    V_next = 0.5*(1 - opts.alpha) * V + Vtilde;
    
    %% get new estimate and calculate corresponding error
    % intermediate rank-2r estimate
    X_hat_2r = U * V_next' + U_next * V' - U * V';
    if opts.normalize_X
        X_hat_2r = X_hat_2r * sig_avg_est;
    end
    
    % rank r projection of intermediate rank-2r solution
    if opts.r_projection_in_iteration
        [U_r, Sigma_r, V_r] = svds(X_hat_2r,r);
        X_hat = U_r * Sigma_r * V_r';
    else
        X_hat = U_next * V_next';
    end

    % store relRes and update X_hat_2r_best if needed
    all_relRes(iter) = relRes; 
    if relRes < best_relRes
        best_relRes = relRes; 
        X_hat_2r_best = X_hat_2r;
    end
    
    % update U, V and X_hat_diff
    U = U_next;
    V = V_next;
    X_hat_diff = norm(X_hat - X_hat_previous, 'fro') / norm(X_hat, 'fro');
    
    %% report
    if opts.verbose
        fprintf('[INSIDE GNMR] iter %4d \t diff X_r %5d\t relRes %6d\n',...
            iter, X_hat_diff, relRes);
    end

    %% check early stopping criteria
    if relRes < opts.stop_relRes
        msg = '[INSIDE GNMR] Early stopping: small error on observed entries\n';
        convergence_flag = 1;
    elseif X_hat_diff < opts.stop_relDiff
        msg = '[INSIDE GNMR] Early stopping: X_hat does not change\n';
        convergence_flag = 1;
    elseif iter > 1 && ...
            abs(all_relRes(iter-1)/relRes-1) < opts.stop_relResDiff
        msg = '[INSIDE GNMR] Early stopping: relRes does not change\n';
        convergence_flag = 1;
    elseif iter > 1 && LSQR_iters_done == 0
        msg = '[INSIDE GNMR] Early stopping: no iterations of LSQR solver\n';
        convergence_flag = 1;
    elseif iter >= 3 * opts.stop_relResStuck_iters && opts.stop_relResStuck_ratio > 0 ...
            && mod(iter, opts.stop_relResStuck_iters) == 0
        % the factor of 3 is since we want to ignore the first opts.stop_relResStuck_iters
        % iteratios, as they may contain small observed relRes with large unobserved...
        last_relRes = all_relRes(iter-opts.stop_relResStuck_iters+1:iter);
        previous_relRes = all_relRes(iter-2*opts.stop_relResStuck_iters+1:iter-opts.stop_relResStuck_iters);
        if min(last_relRes) / min(previous_relRes) > opts.stop_relResStuck_ratio
            msg = sprintf('[INSIDE GNMR] Early stopping: relRes did not decrease by a factor of %f in %d iterations\n', ...
                opts.stop_relResStuck_ratio, opts.stop_relResStuck_iters);
            convergence_flag = 1;
        end
    end
    if convergence_flag
        if opts.verbose
            fprintf(msg);
        end
        break
    end
    
    %% update before next iterate
    X_hat_previous = X_hat;
end

%% return
[U_hat, lambda_hat, V_hat] = svds(X_hat_2r_best,r);
X_hat = U_hat * lambda_hat * V_hat';
end

%% auxiliary functions for generating sparse matrix A
function A = generate_sparse_A(U,V,omega,colind_A)
    nv = size(omega,1);
    r = size(U,2);
    n1 = size(U,1);
    n2 = size(V,1);
    %A = zeros(nv,m);   % matrix of least squares problem 
    val_A = zeros(1,2*r*nv);
    rowind_A=kron(1:nv,ones(1,2*r));
    for counter=1:nv
        j = omega(counter,1); k = omega(counter,2);
        %colind_A = [colind_A,(r*(k-1)+1):(r*(k-1)+r),(r*nc + r*(j-1)+1):(r*nc + r*(j-1)+r)];
        val_A((counter-1)*(2*r)+1:counter*2*r) = [U(j,:),V(k,:)];
    end
    A = sparse(rowind_A,colind_A,val_A,nv,r*(n1+n2));
end

function colind_A = generate_sparse_matrix_indices(omega,r,n2)
    nv = length(omega);
    colind_A=zeros(1,2*r*nv);
    for counter=1:nv
        j = omega(counter,1); k = omega(counter,2);
        colind_A((counter-1)*(2*r)+1:counter*2*r) = [(r*(k-1)+1):(r*(k-1)+r),...
            (r*n2 + r*(j-1)+1):(r*n2 + r*(j-1)+r)];
    end
end
